### PR TITLE
feat(data-table): Add documentation for component / foundation / adapter

### PR DESCRIPTION
**Pull request by abhiomkar**
Refs #4591

### CODE REVIEW

File: packages/mdc-data-table/README.md [hunk 1]
```diff
@@ -34,6 +34,19 @@ npm install @material/data-table
...
+```js
+import {MDCDataTable} from '@material/data-table';
+const dataTable = new MDCDataTable(document.querySelector('.mdc-data-table'));
+```
+
+> See [Importing the JS component](../../docs/importing-js.md) for more information on how to import JavaScript.
+
+> Instantiating `MDCDataTable` component is only required to add interactions for example, row selection.
+
+MDC Data Table component auto instantiates `MDCCheckbox` for header row checkbox and all row checkbox. Make sure to set required class names to instantiate checkbox component. We suggest to use `layout` API when rows are added or removed from data table to register new checkbox components.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: Can you provide a quick example of what using the `layout` API looks like...or do we have an example of this in another readme? Maybe you can just link to the example.

File: packages/mdc-data-table/README.md [hunk 2]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
 ## Accessibility
 
 Please refer [WAI-ARIA Authoring Practices for table](https://www.w3.org/TR/wai-aria-practices-1.1/#table) for ARIA recommended role, states & properties required for table element.
+
+## Events
+
+Please refer MDCDataTable's constant file to access these event constants.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: refer to

File: packages/mdc-data-table/README.md (hunk 1 above)
Comment (moog16) [inline packages/mdc-data-table/README.md]: `... all row checkboxes...`

File: packages/mdc-data-table/README.md [hunk 3]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
...
+-- | -- | --
+`ROW_SELECTION_CHANGED` | `MDCDataTable:changed` | Event emitted when row checkbox is checked or unchecked.
+`SELECTED_ALL` | `MDCDataTable:selectedAll` | Event emitted when header row checkbox is checked.
+`UNSELECTED_ALL` | `MDCDataTable:unselectedAll` | Event emitted when header row checkbox is unchecked.
+
+## `MDCDataTable` Properties and Methods
+
+Method Signature | Description
+--- | ---
+`layout() => void` | Registers new row checkbox, header row checkbox and updates the state of header row checkbox. Use this when rows are added / removed from data table.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: `...new row checkboxes...`

File: packages/mdc-data-table/README.md [hunk 4]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
...
+`getRows() => HTMLElement[]` | Returns array of row elements.
+`getSelectedRowIds() => Array<string|nul>` | Returns array of selected row ids.
+`setSelectedRowIds(rowIds: string[])` | Sets selected row ids. Overwrites previously selected rows.
+
+## Usage within Web Frameworks
+
+If you are using a JavaScript framework, such as React or Angular, you can create a Dialog for your framework. Depending on your needs, you can use the _Simple Approach: Wrapping MDC Web Vanilla Components_, or the _Advanced Approach: Using Foundations and Adapters_. Please follow the instructions [here](../../docs/integrating-into-frameworks.md).
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: a Dialog?

File: packages/mdc-data-table/README.md [hunk 5]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
...
+--- | ---
+`addClassAtRowIndex(rowIndex: number, cssClasses: string) => void` | Adds a class name to row element at given row index excluding header row.
+`getAttributeAtRowIndex(rowIndex: number, attr: string) => void` | Returns attribute value of row element at given row index excluding header row.
+`getRowCount() => number` | Returns row count excluding header row.
+`getRowElements() => HTMLElement[]` | Returns array of row elements excluding header row.
+`getRowIdAtIndex(rowIndex: number) => string | null` | Returns row id of row element at given row index based on `data-row-id` attribute on row element `tr`.
+`getRowIndexByChildElement(el: Element) => number` | Returns index of row element that contains give child element.
+`getSelectedRowCount() => number` | Returns selected row count.
+`isHeaderRowCheckboxChecked() => boolean` | Returns true if header row checkbox is checked.
+`isRowsSelectable() => boolean` | Returns true if table rows are selectable.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: `...if all table rows...`

File: packages/mdc-data-table/README.md [hunk 6]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
...
+`notifyRowSelectionChanged(data: MDCDataTableRowSelectionChangedEventDetail) => void` | Notifies when row selection is changed.
+`notifySelectedAll() => void` | Notifies when header row is checked.
+`notifyUnselectedAll() => void` | Notifies when header row is unchecked.
+`registerHeaderRowCheckbox() => void` | Initializes header row checkbox. Destroys previous header row checkbox instance if any.
+`registerRowCheckboxes() => void` | Initializes all row checkboxes. Destroys previous row checkbox instances if any. This is usually called when row checkboxes are added or removed from table.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: `...destroys all previous row checkbox instances...`?

File: packages/mdc-data-table/README.md [hunk 7]
```diff
@@ -173,3 +187,66 @@ Mixin | Description
...
+`setRowCheckboxCheckedAtIndex(rowIndex: number, checked: boolean) => void` | Sets row checkbox to checked or unchecked at given row index.
+
+### `MDCDataTableFoundation`
+
+Method Signature | Description
+--- | ---
+`layout() => void` | Re-initializes header row checkbox and row checkboxes when selectable rows are added or removed from table.
+`getRows() => HTMLElement[]` | Returns array of row elements.
+`setSelectedRowIds(rowIds: string[]) => void` | Sets selected row ids. Overwrites previously selected rows.
+`getSelectedRowIds() => Array<string|null>` | Returns array of selected row ids.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: does this actually return null? I thought data-table autogens row-ids, so all rows should have an id.

File: packages/mdc-data-table/adapter.ts [hunk 8]
```diff
@@ -32,17 +32,79 @@
...
+   */
   addClassAtRowIndex(rowIndex: number, cssClasses: string): void;
+
+  /**
+   * Returns attribute value of row element at given row index excluding header row.
+   *
+   * @param rowIndex Index of row element excluding header row.
+   * @param attr Attribute string of row element.
+   */
   getAttributeAtRowIndex(rowIndex: number, attr: string): void;
```
Comment (moog16) [inline packages/mdc-data-table/adapter.ts]: I think this returns a `string`, not void. Also add an `@return...`

File: packages/mdc-data-table/adapter.ts [hunk 9]
```diff
@@ -32,17 +32,79 @@
...
+
+  /**
+   * Returns row count excluding header row.
```
Comment (moog16) [inline packages/mdc-data-table/adapter.ts]: `@return number row count excluding header row.`

File: packages/mdc-data-table/adapter.ts [hunk 10]
```diff
@@ -32,17 +32,79 @@
...
+   */
   getRowCount(): number;
+
+  /**
+   * Returns array of row elements excluding header row.
```
Comment (moog16) [inline packages/mdc-data-table/adapter.ts]: @return HTMLElement[] array of row elements excluding header row.

File: packages/mdc-data-table/adapter.ts [hunk 11]
```diff
@@ -32,17 +32,79 @@
...
+  /**
+   * Returns index of row element that contains give child element.
+   *
+   * @param el Child element of row element.
+   * @return Index of row element.
+   */
   getRowIndexByChildElement(el: Element): number;
+
+  /**
+   *
```
Comment (moog16) [inline packages/mdc-data-table/adapter.ts]: delete extra line

File: packages/mdc-data-table/README.md (hunk 1 above)
Comment (abhiomkar) [reply packages/mdc-data-table/README.md]: I'll create additional screenshot test for this is upcoming PRs.

Comment (abhiomkar) [reply packages/mdc-data-table/README.md]: all row checkboxes?

File: packages/mdc-data-table/README.md (hunk 2 above)
Comment (abhiomkar) [reply packages/mdc-data-table/README.md]: This looks correct?

File: packages/mdc-data-table/README.md (hunk 5 above)
Comment (abhiomkar) [reply packages/mdc-data-table/README.md]: Table is not supposed to have some selectable rows. I think this implies that all rows are selectable when at least one row is selectable.

File: packages/mdc-data-table/README.md (hunk 7 above)
Comment (abhiomkar) [reply packages/mdc-data-table/README.md]: The design decision was changed to make row id optional. This'll include 'null' in the list in absence of row-id data attribute for selected row element.

File: packages/mdc-data-table/adapter.ts (hunk 8 above)
Comment (abhiomkar) [reply packages/mdc-data-table/adapter.ts]: Removed this adapter since unused.

File: packages/mdc-data-table/README.md (hunk 1 above)
Comment (moog16) [reply packages/mdc-data-table/README.md]: ` for header row checkbox and all row checkbox` --> ` for header row checkbox and all row checkboxes`

File: packages/mdc-data-table/README.md (hunk 2 above)
Comment (moog16) [reply packages/mdc-data-table/README.md]: Sorry - I meant: `Please refer MDC...` ---> `Please refer to MDC...`

File: packages/mdc-data-table/README.md (hunk 5 above)
Comment (moog16) [reply packages/mdc-data-table/README.md]: Maybe it should say `Returns true if at least one row is selectable.`

**Comment by abhiomkar** [comment]
watered this branch, it's ready for re-review.

File: packages/mdc-data-table/README.md [hunk 12]
```diff
@@ -173,3 +187,65 @@ Mixin | Description
 ## Accessibility
 
 Please refer [WAI-ARIA Authoring Practices for table](https://www.w3.org/TR/wai-aria-practices-1.1/#table) for ARIA recommended role, states & properties required for table element.
+
+## Events
+
+Please use MDCDataTable's constant file to access these event constants.
```
Comment (moog16) [inline packages/mdc-data-table/README.md]: constant file --> constants file

**Pull request merged**